File: src/main/java/net/digital_alexandria/r__/lexer/Lexer.java
package net.digital_alexandria.r__.lexer;

import net.digital_alexandria.r__.exceptions.ParsingException;

/**
 * @author Simon Dirmeier {@literal mail@simon-dirmeier.net}
 */
public final class Lexer
{

    private final static char END_CHAR = '$';

    private String _text;
    private int _pos;
    private char _currChar;

    private Lexer() {}

    public static Lexer instance()
    {
        return new Lexer();
    }

    public void init(String text)
    {
        _text = text + END_CHAR;
        _pos = 0;
        _currChar = _text.charAt(_pos);
    }

    @SuppressWarnings("unchecked")
    public Token next()
    {
        while (_currChar != '$')
        {
            if (Character.isSpaceChar(_currChar))
                skipWhitespace();
            else if (Character.isDigit(_currChar))
                return new Token(TokenCategory.INTEGER, toInt());
            else if (_currChar == '+')
            {
                incrementPosition();
                return new Token(TokenCategory.PLUS, '+');
            }
            else if (_currChar == '-')
            {
                incrementPosition();
                return new Token(TokenCategory.MINUS, '-');
            }
            else if (_currChar == '*')
            {
                incrementPosition();
                return new Token(TokenCategory.MULT, '*');
            }
            else if (_currChar == '/')
            {
                incrementPosition();
                return new Token(TokenCategory.DIV, '/');
            }
            throw new ParsingException("Could not take next token.");
        }
        return new Token(TokenCategory.EOF, -1);
    }

    private void incrementPosition()
    {
        _pos++;
        if (this._pos > _text.length() - 1)
            _currChar = '$';
        else
            _currChar = this._text.charAt(this._pos);
    }

    private Integer toInt()
    {
        StringBuilder res = new StringBuilder();
        while (_currChar != '$' && Character.isDigit(_currChar))
        {
            res.append(_currChar);
            incrementPosition();
        }
        return Integer.parseInt(res.toString());
    }

    private void skipWhitespace()
    {
        while (_currChar != '$' && _currChar == ' ')
            incrementPosition();
    }


}

File: src/lexer.cpp

#include <cctype>

#include "lexer.hpp"
#include "node.hpp"
#include "parsing_exception.hpp"
#include "token_category.hpp"

const char lexer::END_CHAR = '$';

void lexer::init(const std::string& text) const
{
    text_ = text + END_CHAR;
    position_ = 0;
    current_char_ = text_[position_];
}

token lexer::next()  const
{
    while (current_char_ != '$')
    {
        if (current_char_ == ' ')
        {
            skip_whitespace();
            continue;
        }
        if (isdigit(current_char_))
            return token(token_category::INTEGER, node<int>(to_int()));
        else if (current_char_ == '+')
        {
            increment_position();
            return token(token_category::PLUS, node<char>('+'));
        }
        else if (current_char_ == '-')
        {
            increment_position();
            return token(token_category::MINUS, node<char>('-'));
        }
        else if (current_char_ == '*')
        {
            increment_position();
            return token(token_category::MULT, node<char>('*'));
        }
        else if (current_char_ == '/')
        {
            increment_position();
            return token(token_category::DIV, node<char>('/'));
        }
        else if (current_char_ == '(')
        {
            increment_position();
            return token(token_category::LPARENS, node<char>('('));
        }
        else if (current_char_ == ')')
        {
            increment_position();
            return token(token_category::RPARENS, node<char>(')'));
        }
        throw parsing_exception("Could not take next token.");
    }
    return token(token_category::END, node<int>(-1));
}

void lexer::skip_whitespace()  const
{
    while (current_char_ != '$' && current_char_ == ' ')
        increment_position();
}

void lexer::increment_position()  const
{
    position_++;
    if (position_ > text_.size() - 1)
        current_char_ = '$';
    else
        current_char_ = text_[position_];
}

int lexer::to_int()  const
{
    std::string res = "";
    while (current_char_ != '$' && isdigit(current_char_))
    {
        res += current_char_;
        increment_position();
    }
    return static_cast<int>(res);
}